lib/oppen/print_stack.rb

```diff
@@ -0,0 +1,201 @@
+# frozen_string_literal: true
+
+# Oppen.
+module Oppen
+  # Class that reprents a stack that builds an output string
+  # using the values of the tokens that were pushed into it.
+  class PrintStack
+    # Class that represents an item in the print stack.
+    class PrintStackEntry
+      # @return [Integer] Indentation level.
+      attr_reader :offset
+      # @return [Token::BreakType] (Called break in the original paper).
+      attr_reader :break_type
+
+      def initialize(offset, break_type)
+        @offset = offset
+        @break_type = break_type
+      end
+    end
+
+    # @return [Integer] Current available space (Called index in the original paper).
+    attr_reader :space
+
+    def initialize(line_width)
+      # Array representing the stack of PrintStackEntries.
+      @items = []
+
+      # Current available space (Called index in the original paper).
+      @space = line_width
+
+      # Page margin (Called length in the original paper).
+      @margin = line_width
+
+      # IO element that builds the output.
+      @output = StringIO.new
+    end
+
+    # Returns the output of the print stack
+    #
+    # @return [StringIO]
+    def output
+      @output.string
+    end
+
+    # Core method responsible for building the print stack and the output string.
+    #
+    # Called Print in the original paper.
+    #
+    # @param token [Token]
+    # @param token_length [Integer]
+    #
+    # @return [Nil]
+    def print_(token, token_length)
+      case token
+      in Token::Begin
+        handle_begin token, token_length
+      in Token::End
+        handle_end
+      in Token::Break
+        handle_break token, token_length
+      in Token::String
+        handle_string token, token_length
+      end
+    end
+
+    # Handle Begin Token.
+    # @see Token::Begin
+    #
+    # @param token [Token]
+    # @param token_length [Integer]
+    #
+    # @return [Nil]
```

Comment:
And speak of `token`, and `token_length`. Can't we do `token` and `token.length`, so we avoid sending the second param altogether? If so, just create an issue, and we address it later.

Reply:
No we cannot remove the token_length, since its not really the length of the token that is passed as the argument (maybe we should change the name). In the original paper it is named l. I think we should change it to just length